Add configurable RPC timeouts per backend.

jsha opened the pull request:
In the process, break out AMQP config into its own struct, one per service.
The AMQPConfig struct is included by composition in the config structs that need
it. If any given service lacks an AMQP config of its own, it gets a default
value from the top-level AMQP config struct, for deployability reasons.

Tightens the RPC code to take a specific AMQP config, not an over-broad
cmd.Config.

Shortens construction of specific RPC clients so they instatiate the generic
client connection themselves, simplifying per-service startup code.

Remove unused SetTimeout method on RPC clients.

Fixes #1052

jmhodges commented:
LGTM

Ticket for config file update, too, later

bifurcation commented on cmd/activity-monitor/main.go:
```diff
@@ -115,7 +115,8 @@ func main() {
 
-		ch, err := rpc.AmqpChannel(c)
+		amqpConf := c.ActivityMonitor.AMQP
+		ch, err := rpc.AmqpChannel(amqpConf)
```
Seems like it would be cleaner and not that verbose to just make the call directly.

```
ch, err := rpc.AmqpChannel(c.Activitymonitori.AMQP)
```

And likewise below.

bifurcation commented on cmd/admin-revoker/main.go:
```diff
@@ -42,22 +42,18 @@ func setupContext(context *cli.Context) (rpc.RegistrationAuthorityClient, *blog.
-
-	rac, err := rpc.NewRegistrationAuthorityClient(raRPC)
+	amqpConf := c.Revoker.AMQP
+	clientName := "AdminRevoker"
```
It seems like these names should be constants, rather than inline definitions.

bifurcation commented on cmd/boulder-publisher/main.go:
```diff
@@ -23,19 +23,17 @@ func main() {
-
-		sac, err := rpc.NewStorageAuthorityClient(saRPC)
+		pubConf := c.Publisher
+		amqpConf := pubConf.AMQP
```
`amqpConf := c.Publisher.AMQP` and `c.Publisher.MaxConcurrentRPCServerRequests` below.

bifurcation commented on cmd/boulder-sa/main.go:
```diff
@@ -17,9 +17,10 @@ import (
 	app.Action = func(c cmd.Config, stats statsd.Statter, auditlogger *blog.AuditLogger) {
-		go cmd.DebugServer(c.SA.DebugAddr)
+		saConf := c.SA
+		go cmd.DebugServer(saConf.DebugAddr)
```
I really think that making all these new variables impedes readability.

bifurcation commented on cmd/config.go:
```diff
@@ -197,10 +172,37 @@ type Config struct {
+type AMQPConfig struct {
+	Server            string
+	Insecure          bool
+	RA                *RPCServerConfig
```
It seems kind of circular to have `c.VA.AMQP.VA`.  Is there a way we can flatten this out, or use names that make clear what the distinction is?

bifurcation replied:
I gather that the idea here is that each entity has entries here for its depedecies. For example, the `WFE` one would have entries for `RA` and `SA`.  That seems fine for individually configuring timeouts, but I don't think we should put the server names in there.  In addition to being duplicative, you risk having a mismatch between the queue that a component is configured to listen on and the one that its clients are sending messages to.

jsha commented:
Updated this to add consts, and to separate out a ServiceQueue config variable, so we don't need to do things like c.VA.AMQP.VA. This also simplified the constructor for NewAmqpServer!

> That seems fine for individually configuring timeouts, but I don't think we should put the server names in there.

We're trying to move the config more to a model where there's one config file per component, rather than a monolithic config for all components. Right now that means moving config variables into per-service structs, but in the long run it will mean actually separating the files. So I think having the WFE explicitly know about the queue name for the RA makes sense, even though it's possible for the RA to listen on a different queue.

jmhodges commented:
Whoops, conflicts!

jmhodges commented:
Also the simplified constructor is doooope

jsha commented:
Conflicts fixed. @bifurcation, have I convinced you that keeping the queue names in each service's config is good (with the self-references now removed)/

bifurcation commented:
Yeah, this looks OK to me.
